packages/explorer/src/helpers/subscriptionPostMessageRelayHelpers.ts

```diff
@@ -0,0 +1,351 @@
+    }
+    assertUnreachable(this.protocol);
+  }
+  onDisconnected(callback: () => void) {
+    if (this.protocol === 'graphql-ws') {
+      return this.graphWsClient.on('closed', callback);
+    }
+    if (this.protocol === 'subscriptions-transport-ws') {
+      return this.transportSubscriptionClient.onDisconnected(callback);
+    }
+    assertUnreachable(this.protocol);
+  }
+
+  request(
+    params: OperationOptions & {
+      query: string;
+      variables: JSONObject;
+      operationName: string | undefined;
+    }
+  ) {
+    return {
+      subscribe: (subscribeParams: Observer<ExecutionResult<JSONObject>>) => {
+        if (this.protocol === 'graphql-ws') {
+          this.unsubscribeFunctions.push(
+            this.graphWsClient.subscribe(params, {
+              ...subscribeParams,
+              next: (data) =>
+                subscribeParams.next?.(data as ExecutionResult<JSONObject>),
+              error: (error) => subscribeParams.error?.(error as Error),
+              complete: () => {},
+            })
+          );
+        }
+        if (this.protocol === 'subscriptions-transport-ws') {
+          return this.transportSubscriptionClient
+            .request(params)
+            .subscribe(subscribeParams);
+        } else {
+          return undefined;
+        }
+      },
+    };
+  }
+
+  unsubscribeAll() {
+    if (this.protocol === 'graphql-ws') {
+      this.unsubscribeFunctions.forEach((off) => {
+        off();
+      });
+      this.unsubscribeFunctions = [];
+    }
+
+    if (this.protocol === 'subscriptions-transport-ws') {
+      return this.transportSubscriptionClient.unsubscribeAll();
+    }
+  }
+}
+
+function setParentSocketError({
+  error,
+  embeddedIFrameElement,
+  embedUrl,
+}: {
+  error: Error | undefined;
+  embeddedIFrameElement: HTMLIFrameElement;
+  embedUrl: string;
+}) {
+  sendPostMessageToEmbed({
+    message: {
+      name: EXPLORER_SET_SOCKET_ERROR,
+      error,
+    },
+    embeddedIFrameElement,
+    embedUrl,
+  });
+}
+
+function setParentSocketStatus({
+  status,
+  embeddedIFrameElement,
+  embedUrl,
+}: {
+  status: SocketStatus;
+  embeddedIFrameElement: HTMLIFrameElement;
+  embedUrl: string;
+}) {
+  sendPostMessageToEmbed({
+    message: {
+      name: EXPLORER_SET_SOCKET_STATUS,
+      status,
+    },
+    embeddedIFrameElement,
+    embedUrl,
+  });
+}
+
+export function executeSubscription({
+  operation,
+  operationName,
+  variables,
+  headers,
+  embeddedIFrameElement,
+  operationId,
+  embedUrl,
+  subscriptionUrl,
+  protocol,
+}: {
+  operation: string;
+  operationId: string;
+  embeddedIFrameElement: HTMLIFrameElement;
+  operationName: string | undefined;
+  variables?: JSONObject;
+  headers?: Record<string, string>;
+  embedUrl: string;
+  subscriptionUrl?: string;
+  protocol: GraphQLSubscriptionLibrary;
+}) {
+  if (!subscriptionUrl) {
+    setParentSocketError({
+      error: new Error('must set subscription URL in embed config'),
```

Comment:
copy change since we don't have this in the config anymore 🙏 please. We can make subscriptionUrl required!